### PR TITLE
#29 Enable Product-Listing

### PR DESCRIPTION
**Pull request by kagaba-etienne**
#### What does this PR do?

- Adds the product listing routes for both vendors and buyers.
- A vendor can view all products in his collection
- A buyer can view all available products at time.
- Adds documentation of these routes in Swagger

#### Description of Task to be completed

> As a Seller I should be able to retrieve a list of all items in my collection
> As a buyer I should be able to retrieve a list of all items on the available list of products
> Such that a high level inspection is viable
> 
> Acceptance criteria
> 
> Given a get request to retrieve all of a seller's items
> Scenario 1
> When the requesting user/token is a seller and the target item already exist in the particular seller's collection and the request body is validated
> Then:
> A json object containing a status code, message, and items as an array should be returned. If no Items, an empty array be returned as the value of the items property
> Scenario 2
> When the request body fails any validation checks or the token is not of the role seller
> Then:
> 
> A json object containing an status property with the appropriate 4XX code, and an error property with the correct error message should be sent to the requesting client describing the validation error

#### How should this be manually tested?

- Make sure you have necessary environment variables (check the config directory)
- First checkout my branch `ft-product-listing-29`
- install necessary dependencies by running `npm install`
- start the server by running `npm run start`
- /!\ Note: All of the following routes are protected, you should access route `i`  with vendor privileges, and route `ii`  with buyer privileges.
- use postman to test these routes:
             **i**.  POST `http://localhost:3000/api/vendor/collection` with a requestBody containing id  property whose value is the id of the vendor requesting his collection.
             **ii.**  GET `http://localhost:3000/api/user/products`
#### What are the relevant pivotal tracker stories?

- [Card #29](https://trello.com/c/yBHTSZAj)

### CODE REVIEW

**Review by Muhet — APPROVED** [review]
Greate job

**Review by ALVINdimpos — APPROVED** [review]

File: src/database/migrations/20230403191535-create-product.cjs [hunk 1]
```diff
@@ -10,15 +10,55 @@ module.exports = {
...
         type: Sequelize.STRING,
+        allowNull: false,
+      },
+      category_id: {
+        type: Sequelize.INTEGER,
+      },
+      picture_urls: {
+        type: Sequelize.ARRAY(Sequelize.STRING),
+<<<<<<< HEAD
+        allowNull: false,
```
Comment (chanceDira) [inline src/database/migrations/20230403191535-create-product.cjs]: Solve this conflict

File: src/database/migrations/20230403191535-create-product.cjs [hunk 2]
```diff
@@ -10,15 +10,55 @@ module.exports = {
...
+        defaultValue: 0,
+      },
+      expiryDate: {
+        type: Sequelize.DATE,
+      },
+      available: {
+        type: Sequelize.BOOLEAN,
+        allowNull: false,
+        defaultValue: false,
+<<<<<<< HEAD
```
Comment (chanceDira) [inline src/database/migrations/20230403191535-create-product.cjs]: conflict

File: src/database/models/cart_items.js [hunk 3]
```diff
@@ -9,21 +9,24 @@ export default (sequelize, DataTypes) => {
...
       });
       Cart_items.belongsTo(models.Product, {
         foreignKey: 'product_id',
         as: 'product',
       });
     }
   }
-  Cart_items.init({
+  Cart_items.init({    User_id: DataTypes.INTEGER,
+<<<<<<< HEAD
```
Comment (chanceDira) [inline src/database/models/cart_items.js]: conflict

File: src/database/models/product.js [hunk 4]
```diff
@@ -1,6 +1,10 @@
-'use strict';
+<<<<<<< HEAD
+import { Model } from 'sequelize';export default (sequelize, DataTypes) => {
+=======
 import { Model } from 'sequelize';
+
 export default (sequelize, DataTypes) => {
+>>>>>>> 9b6b4a2 (feat(product-listing): create the product listing feature)
```
Comment (chanceDira) [inline src/database/models/product.js]: conflict

File: src/database/models/user.js [hunk 5]
```diff
@@ -10,13 +8,15 @@ export default (sequelize, DataTypes) => {
      */
     static associate(models) {
       // define association here
+<<<<<<< HEAD
```
Comment (chanceDira) [inline src/database/models/user.js]: conflict

File: src/database/seeders/20230403202849-Products.cjs [hunk 6]
```diff
@@ -1,20 +1,68 @@
...
-        Name: 'Product 1',
-        description: 'Product 1 description',
-        price: '100',
-        vendor_id: '1',
-      } 
-    ], {});
+        Name: 'Samsung Galaxy S21',
+        description: 'Samsung Galaxy S21 is a smartphone that was tested with the Android 11 operating system. This model weighs 6.2 ounces, has a 6.2-inch touch screen display, 12-megapixel main camera, and 10-megapixel selfie camera. It comes with 8GB of RAM.',
+        price: '1000',
+<<<<<<< HEAD
```
Comment (chanceDira) [inline src/database/seeders/20230403202849-Products.cjs]: conflict

File: src/middleware/roles.js [hunk 7]
```diff
@@ -35,19 +35,18 @@ const isSeller = async (req, res, next) => {
     return res.status(401).json({ message: "Token not provided" }); // assuming the token is sent in the Authorization header
```
Comment (chanceDira) [inline src/middleware/roles.js]: Here you can use "Not authorized", let not use token keyword. some people might not be aware about it :)

**Review by chanceDira — CHANGES_REQUESTED** [review]
Nice work @kagaba-etienne 
Just some changes you need to fix and conflicts

File: src/middleware/roles.js (hunk 7 above)
Comment (kagaba-etienne) [reply src/middleware/roles.js]: ooh, okay it is fine. I'll change it.

**Review by kagaba-etienne — COMMENTED** [review]
Done!

**Review by ALVINdimpos — APPROVED** [review]

**Review by faid-terence — APPROVED** [review]
Good Job @kagaba-etienne

**Review by nili-alu — APPROVED** [review]

**Review by chanceDira — APPROVED** [review]
Nice job @kagaba-etienne

**Pull request merged**